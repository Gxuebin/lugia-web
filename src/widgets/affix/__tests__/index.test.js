/**
 *
 * create by ligx
 *
 * @flow
 */
import React from 'react';
import chai from 'chai';

import 'jest-styled-components';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import AffixDemo from '../demo';
import renderer from 'react-test-renderer';
import Affix from '../affix';

const { expect: exp } = chai;

Enzyme.configure({ adapter: new Adapter() });

describe('Affix', () => {
  it('css', () => {
    const Target = <AffixDemo />;
    expect(renderer.create(Target).toJSON()).toMatchSnapshot();
  });

  const getState = (target: any) => {
    return target.instance().state;
  };
  it('Affix props: offsetTop setFixedForWin ', () => {
    const target = mount(<Affix offsetTop={50} />);
    const affixElement = target.instance();

    expect(getState(target).fixed).toBe(false);

    affixElement.setFixedForWin({
      offsetTop: 50,
      offsetBottom: 0,
      winHeight: 900,
      scrollTop: 100,
      affixOffsetTop: 150,
      defaultOffsetTop: 100,
    });
    expect(getState(target).fixed).toBe(true);

    affixElement.setFixedForWin({
      offsetTop: 50,
      offsetBottom: 0,
      winHeight: 900,
      scrollTop: 100,
      affixOffsetTop: 150,
      defaultOffsetTop: 200,
    });
    expect(getState(target).fixed).toBe(false);
  });

  it('Affix props: offsetBottom setFixedForWin ', () => {
    const target = mount(<Affix offsetBottom={50} />);
    const affixElement = target.instance();

    expect(getState(target).fixed).toBe(false);

    affixElement.affix.getBoundingClientRect = () => {
      return {
        top: 100,
        bottom: 850,
      };
    };
    affixElement.setFixedForWin({
      offsetTop: 0,
      offsetBottom: 50,
      winHeight: 900,
      scrollTop: 100,
      affixOffsetTop: 150,
      defaultOffsetTop: 200,
    });
    expect(getState(target).fixed).toBe(true);

    affixElement.setFixedForWin({
      offsetTop: 50,
      offsetBottom: 0,
      winHeight: 900,
      scrollTop: 100,
      affixOffsetTop: 150,
      defaultOffsetTop: 200,
    });
    expect(getState(target).fixed).toBe(false);
  });

  it('Affix props: targt offsetTop setFixedForTarget ', () => {
    const target = mount(<Affix target={() => {}} offsetTop={50} />);
    const affixElement = target.instance();

    expect(getState(target).fixed).toBe(false);

    affixElement.setFixedForTarget({
      affixRect: {
        top: 350,
        bottom: 400,
      },
      targetRect: {
        top: 200,
        bottom: 300,
      },
      targetScroll: 100,
      offsetTop: 50,
      winHeight: 900,
      offsetBottom: 0,
    });
    expect(getState(target).fixed).toBe(true);

    affixElement.setFixedForTarget({
      affixRect: {
        top: 350,
        bottom: 400,
      },
      targetRect: {
        top: 100,
        bottom: 400,
      },
      targetScroll: 100,
      offsetTop: 50,
      winHeight: 900,
      offsetBottom: 0,
    });
    expect(getState(target).fixed).toBe(true);
  });

  it('Affix props: targt offsetBottom setFixedForTarget ', () => {
    const target = mount(<Affix target={() => {}} offsetBottom={50} />);
    const affixElement = target.instance();

    expect(getState(target).fixed).toBe(false);
    // targetRect.bottom - affixRect.bottom - targetScroll <= offsetBottom ||
    // affixRect.top <= targetRect.top
    affixElement.setFixedForTarget({
      affixRect: {
        top: 350,
        bottom: 550,
      },
      targetRect: {
        top: 200,
        bottom: 300,
      },
      targetScroll: 100,
      offsetTop: 0,
      winHeight: 900,
      offsetBottom: 50,
    });
    expect(getState(target).fixed).toBe(true);

    affixElement.setFixedForTarget({
      affixRect: {
        top: 100,
        bottom: 600,
      },
      targetRect: {
        top: 150,
        bottom: 400,
      },
      targetScroll: 100,
      offsetTop: 50,
      winHeight: 900,
      offsetBottom: 0,
    });
    expect(getState(target).fixed).toBe(true);
  });
});
