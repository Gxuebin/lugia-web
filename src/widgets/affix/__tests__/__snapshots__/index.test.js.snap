// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Affix css 1`] = `
.c0 {
  font-size: 1.2rem;
  top: 0em;
}

.c1 {
  display: inline-block;
  margin-bottom: 0;
  box-sizing: border-box;
  text-align: center;
  touch-action: manipulation;
  cursor: pointer;
  white-space: nowrap;
  line-height: 1;
  font-family: Trebuchet Ms,Arial,Helvetica,sans-serif;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all 0.3s cubic-bezier(0.645,0.045,0.355,1);
  transition: all 0.3s cubic-bezier(0.645,0.045,0.355,1);
  position: relative;
  text-transform: none;
  outline: 0;
  color: #666666;
  background-color: #ffffff;
  border: 1px solid #cccccc;
  padding: 0 1.2857142857142858em;
  font-size: 1.4rem;
  border-radius: 0.28571428571428575em;
}

.c1:hover {
  color: #9482ff;
  border: 1px solid #9482ff;
  background-color: rgba(104,79,255,0.05);
}

.c1:focus {
  color: #9482ff;
  border: 1px solid #9482ff;
  background-color: rgba(104,79,255,0.05);
}

.c1:active {
  color: #533fcc;
  border: 1px solid #533fcc;
}

.c2 {
  display: inline-block;
  line-height: 2.142857142857143em;
}

<div>
  <div
    style={
      Object {
        "height": "1600px",
        "position": "relative",
        "width": "200px",
      }
    }
  >
    <div
      style={
        Object {
          "height": "200px",
          "width": "200px",
        }
      }
    />
    <div
      style={null}
    >
      <div
        className="c0"
      >
        <p>
          window: affix-top
        </p>
        <button
          className="c1"
          onClick={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          onMouseUp={[Function]}
          size="default"
        >
          <span
            className="c2"
            size="default"
          >
            Affix-top-50
          </span>
        </button>
      </div>
    </div>
    <div
      style={
        Object {
          "height": "200px",
          "overflowY": "scroll",
          "width": "200px",
        }
      }
    >
      <div
        style={
          Object {
            "height": "400px",
          }
        }
      >
        <div
          style={
            Object {
              "height": "100px",
              "width": "20px",
            }
          }
        />
        <div
          style={null}
        >
          <div
            className="c0"
          >
            <p>
              target: target-top
            </p>
            <button
              className="c1"
              onClick={[Function]}
              onMouseDown={[Function]}
              onMouseEnter={[Function]}
              onMouseOut={[Function]}
              onMouseOver={[Function]}
              onMouseUp={[Function]}
              size="default"
            >
              <span
                className="c2"
                size="default"
              >
                target-top-50
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      style={
        Object {
          "height": "200px",
          "overflowY": "scroll",
          "position": "absolute",
          "top": "800px ",
          "width": "200px",
        }
      }
    >
      <div
        style={
          Object {
            "height": "400px",
          }
        }
      >
        <div
          style={
            Object {
              "height": "200px",
              "width": "20px",
            }
          }
        />
        <div
          style={null}
        >
          <div
            className="c0"
          >
            <p>
              target: target-bottom
            </p>
            <button
              className="c1"
              onClick={[Function]}
              onMouseDown={[Function]}
              onMouseEnter={[Function]}
              onMouseOut={[Function]}
              onMouseOver={[Function]}
              onMouseUp={[Function]}
              size="default"
            >
              <span
                className="c2"
                size="default"
              >
                target-bottom-50
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      style={
        Object {
          "height": "600px",
          "width": "200px",
        }
      }
    />
    <div
      style={null}
    >
      <div
        className="c0"
      >
        <p>
          window: affix-bottom
        </p>
        <button
          className="c1"
          onClick={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          onMouseUp={[Function]}
          size="default"
        >
          <span
            className="c2"
            size="default"
          >
            Affix-bottom-50
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
